packages/typescript-estree/src/parser.ts

```diff
@@ -468,7 +527,33 @@ function parseAndGenerateServices<T extends TSESTreeOptions = TSESTreeOptions>(
   warnAboutTSVersion();
 
   /**
-   * Generate a full ts.Program or offer provided instance in order to be able to provide parser services, such as type-checking
+   * Figure out whether this is a single run or part of a long-running process
+   */
+  inferSingleRun();
+
+  /**
+   * If this is a single run in which the user has not provided any existing programs but there
+   * are programs which need to be created from the provided "project" option,
+   * create the programs once ahead of time and avoid watch programs
+   */
+  if (extra.singleRun && !extra.programs && extra.projects?.length > 0) {
+    extra.programs = extra.projects.map(configFile => {
+      const existingProgram = existingPrograms.get(configFile);
+      if (existingProgram) {
+        return existingProgram;
+      }
+      log(
+        'Detected single-run/CLI usage, creating Program once ahead of time for project: %s',
+        configFile,
+      );
+      const newProgram = createProgramFromConfigFile(configFile);
+      existingPrograms.set(configFile, newProgram);
+      return newProgram;
+    });
```

Comment:
If you run `eslint ./package/eslint-plugin/src/index.ts`, we don't want to load up the programs for every single project just to lint one file.

Right now we will iterate through the `project`s array one at a time, creating watch programs until we find one that matches.

Perhaps if we change the type of `programs` to be an `Iterable`, and then we can store a generator in `extras.program` and just iterate over the iterable?

If you pass in `parserOptions.program` as an array then it still works as well.

clarifying example:
```ts
const projectsIterator = extra.projects[Symbol.iterator];
const programsIterator = {
  next: () => {
    const configFile = projectsIterator.next();
	// calculate the program for the project
  },
  [Symbol.iterator]: () => iterator,
};
extras.programs = programsIterator;
```

Reply:
Yeah this is a good optimization for those use-cases, thanks, I will update it